src/Npgsql/NpgsqlParameterCollection.cs

```
@@ -325,24 +321,14 @@ public override int IndexOf(string parameterName)
         public override void RemoveAt(int index)
         {
             if (_internalList.Count - 1 < index)
-                throw new IndexOutOfRangeException();
+                throw new ArgumentOutOfRangeException(nameof(index));
+
             Remove(_internalList[index]);
         }
 
         /// <inheritdoc />
         public override void Insert(int index, object value)
-        {
-            if (value == null)
-                throw new ArgumentNullException(nameof(value));
-            if (!(value is NpgsqlParameter param))
-                throw new InvalidCastException($"{nameof(value)} must be an NpgsqlParameter");
-            if (param.Collection != null)
-                throw new InvalidOperationException("The parameter already belongs to a collection");
-
-            param.Collection = this;
-            _internalList.Insert(index, param);
-            InvalidateHashLookups();
-        }
+            => Insert(index, (NpgsqlParameter)value);
```

Maybe throw an informative InvalidCastException if value isn't an NpgsqlParameter.